### PR TITLE
Updated version for godot 4.0.2.stable

### PR DESCRIPTION
**Pull request by Hamzalopode**
Hi, I updated the addons to work with 4.0.2.stable, need to be verified but the only issue I'm currently having is the display of yellow line ( outline of the box in the editor ). beside this should be working fine. Let me know if there is anything I will do the change so we can release this for godot 4.0.2.stable.

### CODE REVIEW

**Review by Cryszon — COMMENTED** [review]
Releasing Godot 4 version should be a new major release and we also need to retain the 3.x version (probably in a separate branch). I removed version bumps from this pull request so we can do the release separately after everything is working.

File: addons/level_block/level_block_gizmo.gd
Comment (OneginIII) [inline addons/level_block/level_block_gizmo.gd:1]: Updated this to reflect Godot 4 API changes.

File: addons/level_block/level_block_node.gd [hunk 1]
```diff
@@ -1,61 +1,56 @@
...
 
 signal texture_updated(new_texture)
 signal texture_size_updated(new_size)
 
 const size = 1.0
 const occluder_multiplier = 1.001
 
-export(SpatialMaterial) var material = load("res://addons/level_block/default_material.tres")
-export(Texture) var texture_sheet setget set_texture
+@export var material := load("res://addons/level_block/default_material.tres")
```
Comment (OneginIII) [inline addons/level_block/level_block_node.gd]: Set the material type.

File: addons/level_block/level_block_node.gd [hunk 2]
```diff
@@ -68,7 +63,7 @@ var shape
 var occluders : Array
 
 func _ready():
-	set_notify_transform(true)
+	set_notify_transform(false)
```
Comment (OneginIII) [inline addons/level_block/level_block_node.gd]: This should be true to update transforms when edited in the editor.

File: addons/level_block/level_block_node.gd [hunk 3]
```diff
@@ -81,34 +76,34 @@ func refresh():
...
 	if !generate_collision:
 		return
-	body = PhysicsServer.body_create(PhysicsServer.BODY_MODE_STATIC)
-	PhysicsServer.body_set_mode(body, PhysicsServer.BODY_MODE_STATIC)
-	shape = PhysicsServer.shape_create(PhysicsServer.SHAPE_CONCAVE_POLYGON)
-	PhysicsServer.shape_set_data(shape, mesh.get_faces())
+	body = PhysicsServer3D.body_create()
+	PhysicsServer3D.body_set_mode(body, PhysicsServer3D.BODY_MODE_STATIC)
+	shape = PhysicsServer3D.area_create()
+	PhysicsServer3D.shape_set_data(shape, mesh.get_faces())
```
Comment (OneginIII) [inline addons/level_block/level_block_node.gd]: The shape generation has changed a bit in Godot 4. The API is not fully documented, but this is how this should work, I believe.

File: addons/level_block/level_block_node.gd [hunk 4]
```diff
@@ -188,8 +183,8 @@ func create_occluders():
 	for i in range(6):
 		if faces[i] < 0:
 			continue
-		var occluder = Occluder.new()
-		occluder.shape = OccluderShapePolygon.new()
+		var occluder = OccluderInstance3D.new()
+		occluder.shape = PolygonOccluder3D.new()
```
Comment (OneginIII) [inline addons/level_block/level_block_node.gd:187]: Didn't test occluders at all.

File: addons/level_block/texture_selector.gd
Comment (OneginIII) [inline addons/level_block/texture_selector.gd:1]: The Inspector plugin that shows a preview of the texture on each face doesn't work. It needs to be fixed separately.

**Comment by OneginIII** [comment]
I fixed the issues I found related mostly to Godot 4 API changes. A lot of the new API is undocumented, but hopefully I understood enough of how it works now.

The inspector plugin that shows a preview of the texture on each face doesn't work. I couldn't get it to work with any minor changes and it needs to be looked at separately.

**Review by OneginIII — COMMENTED** [review]
Comments on the original pull request.

**Comment by Hamzalopode** [comment]
Hi, thanks both of you for looking into this. As @Cryszon said this need to be in a separate branch (probably ?), I'll look into the changes both of you made to see if I can be useful.

**Comment by Hamzalopode** [comment]
> I think the biggest issue right now is the custom inspector not working and showing texture previews.

Ok, I will look into this tonight, if I manage to get anything to work correctly I will commit it.

**Comment by Hamzalopode** [comment]
I managed to make the custom inspector work, texture previews work and I'm currently trying to make the whole project (shown in the youtube video) to work. 

But I have a blocker here:

_file: player.gd, line 13 to 15_

`func _physics_process(delta):
	if tween.is_active():
		return`

What is the objective of the "**if**" statement ? Thanks.

**Comment by Hamzalopode** [comment]
Hi, this is the result I managed to get. Custom selector working as shown in this screenshot:
image

Unfortunately I couldn't make the project work due to an error with "Tween" located in player.gd (Shouldn't be a problem, it's probably due to Tween not being a Node anymore). 

Also the rotation didn't seems right, probably due to Tweens? 🤔 

test_2

I'm going to commit those changes, and let me know if there is anything wrong.
Note: I didn't document any changed code, I don't know if it's going to be needed (beside the Tween issue).

**Comment by Hamzalopode** [comment]
Whole project converted to 4.0.3, see [commit](https://github.com/Reun-Media/godot-levelblock/pull/8/commits/15e50042053e4dfe1bd7cb5d382bfaf8d7ebc611) for more information.

Current Bug(s) known:
- Player.gd: Tween issue, I don't know why.

**Review by Hamzalopode — COMMENTED** [review]
Fix typo commit message:
- Custom inspector and textures previews seems to work.*

**Comment by OneginIII** [comment]
I fixed the player movement by copying the logic I did for the tutorial. Since I worked on the plugin before making the tutorial, this demo used a more complicated version of the movement than really necessary. The tutorial still used Godot 3's SceneTreeTween, which is similar to the new Tween system.

I also fixed the test scene visuals and fixed toggling the different draw modes using the T key. This let me test the occlusion culling and I got it working pretty easily. I thought that you had to separately bake the OccluderInstance3Ds, which I didn't find a way to do in code. It seems to just work the way it is now though.

The texture preview for the faces still doesn't work for me.

This is the original 3.x plugin:
Godot_v3 5 1-stable_mono_win64_sq4vhyZc13
This is what it looks like for me in Godot 4:
Godot_v4 0 3-stable_win64_7z2PA76GEl

After that is fixed, we should be ready to release the port.

One very minor thing to fix is removing the tileset texture from the default material or just including it in the addon folder. Since it's not there now, if someone downloads just the addon, the default material will give an error for the missing texture. Need to decide what to do with this, the texture isn't big, so maybe just including it in the addon folder would be fine.

**Comment by OneginIII** [comment]
Alright, I got the inspector plugin working in Godot 4:
Godot_v4 0 3-stable_win64_tlrcXYDCdi

I decided to include the tileset texture in the addon as an example file, since it's only 9 KB in size.

The plugin should now work for Godot 4. Could you @Cryszon help with the changelog and branching?

**Review by OneginIII — APPROVED** [review]
Plugin seems good to go for Godot 4 now. Thanks for your help @Hamzalopode!

**Comment by Hamzalopode** [comment]
> Plugin seems good to go for Godot 4 now. Thanks for your help @Hamzalopode!

Thanks for both of you too!

**Review by Cryszon — APPROVED** [review]

**Pull request merged**